src/merge.rs

```diff
@@ -1792,66 +1759,124 @@ impl<'t> MergedRoot<'t> {
     }
 
     /// Returns a sequence of completion operations, or "completion ops", to
-    /// apply to the local tree so that it matches the merged tree.
-    pub fn completion_ops(&self) -> CompletionOps<'_> {
+    /// apply to the local tree so that it matches the merged tree. The abort
+    /// signal can be used to interrupt fetching the ops.
+    pub fn completion_ops_with_signal(
+        &self,
+        signal: &impl AbortSignal,
+    ) -> Result<CompletionOps<'_>> {
         let mut ops = CompletionOps::default();
-        accumulate(&mut ops, self.node(), 1, false);
-        for guid in self.delete_locally.iter() {
-            if self.local_tree.mentions(guid) && self.remote_tree.mentions(guid) {
-                // Only flag tombstones for items that exist in both trees,
-                // excluding ones for items that don't exist in the local
-                // tree. This can be removed once we persist tombstones in
-                // bug 1343103.
-                let flag_as_merged = FlagAsMerged(guid);
-                ops.flag_as_merged.push(flag_as_merged);
+        accumulate(signal, &mut ops, self.node(), 1, false)?;
+
+        // Clean up tombstones for local and remote items that are revived on
```

Comment:
These works like this:

* Any local tombstones that aren't in `delete_remotely` are removed. They're for revived (deleted locally, but changed remotely) items.
* Any tombstones on both sides (if you delete the same bookmark on two devices while they're both offline, then sync) are removed and flagged as remotely merged.
* Any remote tombstones for items that aren't in `delete_locally`, _and that don't exist locally_, are flagged as remotely merged. The extra check avoids a redundant `UPDATE items SET needsMerge = 1` for the remote tombstone. Since we'll be reuploading the local item, anyway, it doesn't make sense to update its `needsMerge` flag, only to replace the whole row later when we write the uploaded item back into the mirror.
* Any accepted deletions for items that exist on both sides must be because the items are invalid or non-syncable (left pane roots and queries). We delete the item locally, insert a tombstone into `moz_bookmarks_deleted` in case upload is interrupted, and flag the tombstone for upload.
* Any deletions for items that only exist locally must be either incoming tombstones (`{ id: "xyz", deleted: true }`), or local non-syncable items. Delete them, and also flag the remote tombstone as merged if one exists.
* Any deletions for items that only exist remotely must be either tombstones in `moz_bookmarks_deleted` that we're uploading, or deletions of invalid or non-syncable remote items (left pane on the server, bookmarks with invalid URLs, and such). Insert a tombstone into `moz_bookmarks_deleted` if one doesn't already exist, and flag it for upload.